File: lession5/Obserber/java/WithPattern/Extension/src/main/java/ConsoleProgress.java
import java.text.DecimalFormat;

/**
 * Created by Administrator on 2017/1/17.
 */
public class ConsoleProgress implements Progress {
    public void update(State state) {
        if (state.isCompleted()) {
            System.out.println("download has complete");
        } else {
            System.out.println(String.format("%s/%s %s%%",
                    state.getRead(), state.getTotal(),
                    formatDouble((double) state.getRead() / state.getTotal() * 100)));
        }
    }

    public static String formatDouble(double d) {
        DecimalFormat df = new DecimalFormat("#");
        return df.format(d);
    }
}

File: lession5/Obserber/csharp/WithPattern/ConsoleProgress/ConsoleProgress/ConsoleProgress.cs
using Protocol;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ConsoleProgress
{
    public class ConsoleProgress : Progress
    {
        public void Update(State state)
        {
            if (state.Completed)
            {
                Console.WriteLine("donwload has completed");
            }
            else
            {
                Console.WriteLine(string.Format("{0}/{1} {2}%", state.Read, state.Total, ((double)state.Read / state.Total * 100).ToString("0")));
            }
        }
    }
}
